WebUI: Support editing tracker tier

Piccirello opened the pull request:
This PR adds the ability to direct modify a tracker's tier from the WebUI. This process is notably different than the GUI, which provides arrows for increasing/decreasing a tracker's tier.

Closes #12233

![Screenshot 2025-07-05 at 06 13 52](https://github.com/user-attachments/assets/a1982c84-ca07-45c6-bedd-a42e57b72aa6)

Piccirello commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1129,17 +1129,23 @@ void TorrentsController::editTrackerAction()
         throw APIError(APIErrorType::NotFound);
 
     const QUrl origTrackerUrl {origUrl};
     const QUrl newTrackerUrl {newUrl};
-    if (origTrackerUrl == newTrackerUrl)
+    const bool isTrackerUrlChanged = origTrackerUrl != newTrackerUrl;
+    const bool isNewTierSpecified = newTier >= 0;
+    if (!isTrackerUrlChanged && !isNewTierSpecified)
         return;
```
This endpoint has logic to return early if the URL is unchanged. With the addition of the `tier` param, some effort was required to maintain this functionality.

The logic here still returns early if the URL is unchanged **and** the tier is unspecified. If the tier is specified, the logic for returning early is moved to the loop, where it checks that both the URL and tier are unchanged before returning early.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1129,17 +1129,23 @@ void TorrentsController::editTrackerAction()
     const auto id = BitTorrent::TorrentID::fromString(params()[u"hash"_s]);
     const QString origUrl = params()[u"origUrl"_s];
     const QString newUrl = params()[u"newUrl"_s];
+    // min tier is 0, so -1 indicates to keep tier unchanged
```
Why not return "Bad request" in case of any invalid value is specified?

glassez replied:
I was referring to the inconsistent behavior of the current code. It returns an error for some invalid values (numbers < -1 and > 255), but silently ignores others (number -1 and any other non-numeric garbage).

Piccirello replied:
The code now correctly errors for non-integers and values less than 0.

glassez commented:
I would expect that "newUrl" parameter becomes optional with this change. Otherwise, the client must set it even if they only want to change the tier.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1125,22 +1125,44 @@ void TorrentsController::addTrackersAction()
+    {
+        bool ok = false;
+        const auto number = newTierParam.value().toInt(&ok);
+        if (!ok)
+            throw APIError(APIErrorType::BadParams, u"New tier must be an integer"_s);
+        if ((number < 0) || (number > 255))
+            throw APIError(APIErrorType::BadParams, u"New tier must be between 0 and 255"_s);
```
IMO, "New" is redundant in those sentences.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1125,22 +1125,44 @@ void TorrentsController::addTrackersAction()
+        const QUrl url = newUrlParam.value();
+        if (!url.isValid())
+            throw APIError(APIErrorType::BadParams, u"New tracker URL is invalid"_s);
+        if (url != origTrackerUrl)
```
Actually we can consider the case of `newUrl == url` as a "bad params", no?

Piccirello replied:
Potentially yes though when using this form in the WebUI, it seems strange that if you open the Edit Tracker dialog, and then click Save without changing anything, it displays an error. I think it makes sense to just consider that a no-op, which would mean the API returning a 200/204.

Piccirello replied:
I think it's more common for an "edit" endpoint to treat a lack of changes as a no-op, rather than an explicit error.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1125,22 +1125,44 @@ void TorrentsController::addTrackersAction()
+            newTrackerUrl = url;
+    }
+
+    if (!newTrackerUrl && !newTier)
```
Given the above comment there shouldn't be the case.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1151,7 +1173,7 @@ void TorrentsController::editTrackerAction()
         const QUrl trackerUrl {tracker.url};
 
-        if (trackerUrl == newTrackerUrl)
+        if (newTrackerUrl && (trackerUrl == newTrackerUrl.value()))
```
You can compare `optional` with its value type:
```suggestion
        if (newTrackerUrl && (trackerUrl == newTrackerUrl))
```

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1162,8 +1184,15 @@ void TorrentsController::editTrackerAction()
 
         if (trackerUrl == origTrackerUrl)
         {
+            const bool isTrackerTierChanged = newTier && (tracker.tier != newTier.value());
```
```suggestion
            const bool isTrackerTierChanged = newTier && (tracker.tier != newTier);
```

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1125,22 +1125,44 @@ void TorrentsController::addTrackersAction()
+    }
+
+    if (!newTrackerUrl && !newTier)
         return;
```
To make this endpoint behavior consistent you seem should move `setResult(QString());` above the first `return;` statement.

Piccirello replied:
You're right it's inconsistent. I'm going to change this API now to instead always return a 204 in these cases.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1125,22 +1125,44 @@ void TorrentsController::addTrackersAction()
     if (!torrent)
         throw APIError(APIErrorType::NotFound);
+    if (!newUrlParam && !newTierParam)
+        throw APIError(APIErrorType::BadParams, u"Must specify at least one of [newUrl, tier]"_s);
```
It should be translatable. Please check other strings too.
```suggestion
        throw APIError(APIErrorType::BadParams, tr("Must specify at least one of [newUrl, tier]"));
```

Chocobo1 commented on WebAPI_Changelog.md:
```diff
@@ -1,5 +1,11 @@
 # WebAPI Changelog
 
+## 2.13.0
```
webapplication.h needs to be updated too.

Piccirello commented:
Addressed feedback in https://github.com/qbittorrent/qBittorrent/compare/a2589b28bd7da69f0bed877685be7eb38f3d12f7..8a7db51fe831e341c7b3657a6fc53f1c50df38ca

Piccirello commented:
Force pushed to fix conflict in `WebAPI_Changelog.md`

Chocobo1 commented:
@Piccirello 
Thank you!
